File: src/inner.f90
!-----------------------------------------------------------------------
!
! MODULE: inner_module
!> @brief
!> This module controls the inner iterations. Inner iterations include
!> the KBA mesh sweep, which is parallelized via MPI and vectorized over
!> angles in a given octant. Inner source computed here and inner
!> convergence is checked.
!
!-----------------------------------------------------------------------

MODULE inner_module

  USE global_module, ONLY: i_knd, r_knd, one, ounit, zero

  USE geom_module, ONLY: nx, ny, nz, nc

  USE sn_module, ONLY: nmom, cmom, lma

  USE data_module, ONLY: ng

  USE control_module, ONLY: epsi, tolr, dfmxi, inrdone, it_det

  USE solvar_module, ONLY: q2grp0, q2grpm, s_xs, flux0, flux0pi, fluxm,&
    qtot

  USE sweep_module, ONLY: sweep

  USE time_module, ONLY: tinrsrc, tsweeps, wtime

  USE plib_module, ONLY: nthreads, glmax, comm_snap, iproc, root, ichunk

  USE thrd_comm_module, ONLY: assign_thrd_set

  IMPLICIT NONE

  PRIVATE

  PUBLIC :: inner


  CONTAINS


  SUBROUTINE inner ( inno, iits, t, do_grp, ng_per_thrd, nnstd_used,   &
    grp_act )

!-----------------------------------------------------------------------
!
! Do a single inner iteration for all groups. Calculate the total source
! for each group and sweep the mesh over octants. Repeat for all groups
! unless the group is converged according to inrdone(g).
!
!-----------------------------------------------------------------------

    INTEGER(i_knd), INTENT(IN) :: inno, t

    INTEGER(i_knd), INTENT(INOUT) :: ng_per_thrd, nnstd_used

    INTEGER(i_knd), DIMENSION(ng), INTENT(OUT) :: iits

    INTEGER(i_knd), DIMENSION(ng), INTENT(INOUT) :: do_grp

    INTEGER(i_knd), DIMENSION(ng,nthreads), INTENT(INOUT) :: grp_act
!_______________________________________________________________________
!
!   Local variables
!_______________________________________________________________________

    INTEGER(i_knd) :: n, g

    REAL(r_knd) :: t1, t2, t3, t4
!_______________________________________________________________________
!
!   Compute the inner source and add it to fixed + out-of-group sources.
!   No need to do inner operations if the group's inners are converged.
!_______________________________________________________________________

  !$OMP MASTER

    CALL wtime ( t1 )

    do_grp = 1
    WHERE( inrdone ) do_grp = 0

    CALL assign_thrd_set ( do_grp, ng, ng_per_thrd, ny*nz, nnstd_used, &
      grp_act )

  !$OMP END MASTER
  !$OMP BARRIER

    CALL inner_src ( ng_per_thrd, nnstd_used, grp_act(:,t) )
  !$OMP BARRIER

  !$OMP MASTER
    CALL wtime ( t2 )
    tinrsrc = tinrsrc + t2 - t1
  !$OMP END MASTER
!_______________________________________________________________________
!
!   With source computed, set previous copy of flux; new flux moments
!   iterates computed during sweep.
!_______________________________________________________________________

    DO n = 1, ng_per_thrd
      g = grp_act(n,t)
      IF ( g == 0 ) EXIT
      flux0pi(:,:,:,g) = flux0(:,:,:,g)
    END DO
!_______________________________________________________________________
!
!   Call for the transport sweep. Check convergence, using threads.
!_______________________________________________________________________

  !$OMP MASTER
    CALL wtime ( t3 )
  !$OMP END MASTER

  !$OMP BARRIER

    CALL sweep ( t, do_grp, ng_per_thrd, nnstd_used, grp_act )
  !$OMP BARRIER

  !$OMP MASTER

    CALL wtime ( t4 )
    tsweeps = tsweeps + t4 - t3
!_______________________________________________________________________
!
!   Check inner convergence. Apply nested threads to group sets.
!_______________________________________________________________________

    do_grp = 1
    WHERE( inrdone ) do_grp = 0

    CALL assign_thrd_set ( do_grp, ng, ng_per_thrd, 0, nnstd_used,     &
      grp_act )

  !$OMP END MASTER
  !$OMP BARRIER

    CALL inner_conv ( inno, iits, ng_per_thrd, nnstd_used,             &
      grp_act(:,t) )
!_______________________________________________________________________
!_______________________________________________________________________

  END SUBROUTINE inner


  SUBROUTINE inner_src ( ng_per_thrd, nnstd_used, grp_act )

!-----------------------------------------------------------------------
!
! Compute the inner source, i.e., the within-group scattering source.
! Thread over groups and use nested threads on i-lines when available.
!
!-----------------------------------------------------------------------

    INTEGER(i_knd), INTENT(IN) :: ng_per_thrd, nnstd_used

    INTEGER(i_knd), DIMENSION(ng), INTENT(IN) :: grp_act
!_______________________________________________________________________
!
!   Local variables
!_______________________________________________________________________

    INTEGER(i_knd) :: n, g, k, j
!_______________________________________________________________________
!
!   Loop over each set of groups. Use nested threads if available.
!_______________________________________________________________________

  !$OMP PARALLEL NUM_THREADS(nnstd_used) IF(nnstd_used>1)              &
  !$OMP& DEFAULT(SHARED) PRIVATE(n,g,k,j) PROC_BIND(CLOSE)
    DO n = 1, ng_per_thrd

      g = grp_act(n)
      IF ( g == 0 ) EXIT

  !$OMP DO SCHEDULE(STATIC,1) COLLAPSE(2)
      DO k = 1, nz
      DO j = 1, ny
        CALL inner_src_calc ( j, k, s_xs(:,:,:,:,g), flux0(:,j,k,g),   &
          fluxm(:,:,j,k,g), q2grp0(:,j,k,g), q2grpm(:,:,j,k,g),        &
          qtot(:,:,:,:,:,g) )
      END DO
      END DO
  !$OMP END DO NOWAIT

    END DO
  !$OMP END PARALLEL
!_______________________________________________________________________
!_______________________________________________________________________

  END SUBROUTINE inner_src


  SUBROUTINE inner_src_calc ( j, k, sxs_g, f0, fm, qo0, qom, q )

!-----------------------------------------------------------------------
!
! Compute the within-group scattering for a given group. Add it to fixed
! and out-of-group sources.
!
!-----------------------------------------------------------------------

    INTEGER(i_knd), INTENT(IN) :: j, k

    REAL(r_knd), DIMENSION(nx), INTENT(IN) :: f0, qo0

    REAL(r_knd), DIMENSION(cmom-1,nx), INTENT(IN) :: fm, qom

    REAL(r_knd), DIMENSION(nx,ny,nz,nmom), INTENT(IN) :: sxs_g

    REAL(r_knd), DIMENSION(cmom,ichunk,ny,nz,nc), INTENT(OUT) :: q
!_______________________________________________________________________
!
!   Local variables
!_______________________________________________________________________

    INTEGER(i_knd) :: ic, ich, i, mom, l
!_______________________________________________________________________
!
!   Loop over i cells. Set the first source moment with flux0 (f0). Then
!   set remaining source moments with fluxm (fm) and combination of
!   higher scattering orders.
!_______________________________________________________________________

    ic = 0
    ich = 1

    DO i = 1, nx

      ic = ic + 1

      q(1,ic,j,k,ich) = qo0(i) + f0(i)*sxs_g(i,j,k,1)

      mom = 1
      DO l = 2, nmom
        q(mom+1:mom+lma(l),ic,j,k,ich) = qom(mom:mom+lma(l)-1,i) +     &
          fm(mom:mom+lma(l)-1,i)*sxs_g(i,j,k,l)
        mom = mom + lma(l)
      END DO

      IF ( ic == ichunk ) THEN
        ic = 0
        ich = ich + 1
      END IF

    END DO
!_______________________________________________________________________
!_______________________________________________________________________

  END SUBROUTINE inner_src_calc


  SUBROUTINE inner_conv ( inno, iits, ng_per_thrd, nnstd_used, grp_act )

!-----------------------------------------------------------------------
!
! Check for inner iteration convergence using the flux array.
!
!-----------------------------------------------------------------------

    INTEGER(i_knd), INTENT(IN) :: inno, ng_per_thrd, nnstd_used

    INTEGER(i_knd), DIMENSION(ng), INTENT(OUT) :: iits

    INTEGER(i_knd), DIMENSION(ng), INTENT(IN) :: grp_act
!_______________________________________________________________________
!
!   Local variables
!_______________________________________________________________________

    INTEGER(i_knd) :: n, g

    REAL(r_knd), DIMENSION(nx,ny,nz,ng_per_thrd) :: df
!_______________________________________________________________________
!
!   Thread group loops for computing local difference (df) array.
!   compute max for that group. Need a barrier for the main threads.
!_______________________________________________________________________

  !$OMP PARALLEL DO NUM_THREADS(nnstd_used) IF(nnstd_used>1)           &
  !$OMP& SCHEDULE(STATIC,1) DEFAULT(SHARED) PRIVATE(n,g)               &
  !$OMP& PROC_BIND(CLOSE)
    DO n = 1, ng_per_thrd


      g = grp_act(n)
      IF ( g == 0 ) CYCLE

      iits(g) = inno

      df(:,:,:,n) = one
      WHERE( ABS( flux0pi(:,:,:,g) ) < tolr )
        flux0pi(:,:,:,g) = one
        df(:,:,:,n) = zero
      END WHERE
      df(:,:,:,n) = ABS( flux0(:,:,:,g)/flux0pi(:,:,:,g) - df(:,:,:,n) )

      dfmxi(g) = MAXVAL( df(:,:,:,n) )

    END DO
  !$OMP END PARALLEL DO
  !$OMP BARRIER
!_______________________________________________________________________
!
!   All procs then reduce dfmxi for all groups, determine which groups
!   are converged and print requested info
!_______________________________________________________________________

  !$OMP MASTER

    CALL glmax ( dfmxi, ng, comm_snap )
    WHERE( dfmxi <= epsi ) inrdone = .TRUE.

    IF ( iproc==root .AND. it_det==1 ) THEN
      DO g = 1, ng
        WRITE( ounit, 221 ) g, iits(g), dfmxi(g)
      END DO
    END IF

  !$OMP END MASTER
!_______________________________________________________________________

    221 FORMAT( 4X, 'Group ', I3, 4X, ' Inner ', I5, 4X, ' Dfmxi ',    &
                ES11.4 )
!_______________________________________________________________________
!_______________________________________________________________________

  END SUBROUTINE inner_conv


END MODULE inner_module

File: ports/snap-c/inner.c
/***********************************************************************
 * Module: inner.c
 *
 * This module controls the inner iterations. Inner iterations include
 * the KBA mesh sweep, which is parallelized via MPI and vectorized over
 * angles in a given octant. Inner source computed here and inner
 * convergence is checked.
 ***********************************************************************/
#include "snap.h"

// Define macro to index local variable df(nx,ny,nz,ng)
/* df(nx,ny,nz) */
#ifdef ROWORDER
#define DF_4D(X, Y, Z, G)                       \
    df[ X   * NY * NZ * NG                      \
        + Y * NZ * NG                           \
        + Z * NG                                \
        + G]
#else
#define DF_4D(X, Y, Z, G)                       \
    df[ G   * NZ * NY * NX                      \
        + Z * NY * NX                           \
        + Y * NX                                \
        + X ]
#endif

#define F_3D(X, Y, Z)         FLUX_4D(X, Y, Z, (g-1))
#define FM_4D(CMOM1, X, Y, Z) FLUXM_5D(CMOM1, X, Y, Z, (g-1))
#define QO_4D(CMOM1, X, Y, Z) Q2GRP_5D(CMOM1, X, Y, Z, (g-1))
#define CS_4D(NMOM1, X, Y, Z) S_XS_5D(NMOM1, X, Y, Z, (g-1))
#define Q_4D(CMOM1, X, Y, Z)  QTOT_5D(CMOM1, X, Y, Z, (g-1))


/***********************************************************************
 * Do a single inner iteration for all groups. Calculate the total source
 * for each group and sweep the mesh.
 ***********************************************************************/
void inner ( input_data *input_vars, para_data *para_vars, geom_data *geom_vars,
             time_data *time_vars, data_data *data_vars, sn_data *sn_vars,
             control_data *control_vars, solvar_data *solvar_vars,
             sweep_data *sweep_vars, dim_sweep_data *dim_sweep_vars,
             int inno, int *iits, FILE *fp_out, int *ierr )
{
/***********************************************************************
 * Local variables
 ***********************************************************************/
    int g, k, j, i;

    double t1, t2, t3, t4;

/***********************************************************************
 * Compute the inner source and add it to fixed + out-of-group sources
 ***********************************************************************/
    t1 = wtime();

    inr_src ( input_vars, sn_vars, control_vars, solvar_vars, para_vars );

    t2 = wtime();

    TINRSRC += t2 - t1;

/***********************************************************************
 * With source computed, set previous copy of flux and zero out current
 * copies--new flux moments iterates computed during sweep. Thread
 * over groups.
 ***********************************************************************/
#pragma omp parallel for schedule(dynamic, 1) default(shared)   \
    private(g, k, j, i)
    for ( g = 0; g < NG; g++ )
    {
        if ( !INRDONE_1D(g) )
        {
            for ( k = 0; k < NZ; k++ )
            {
                for ( j = 0; j < NY; j++ )
                {
                    for ( i = 0; i < NX; i++ )
                    {
                        FLUXPI_4D(i,j,k,g) = FLUX_4D(i,j,k,g);
                    }
                }
            }
        }
    }

/***********************************************************************
 * Call for the transport sweep. Check convergence, using threads.
 ***********************************************************************/
    t3 = wtime();

    sweep ( input_vars, para_vars, geom_vars, sn_vars, data_vars,
            control_vars, solvar_vars, sweep_vars, dim_sweep_vars, ierr);

    t4 = wtime();

    TSWEEPS += t4 - t3;

    inr_conv ( input_vars, para_vars, control_vars, solvar_vars,
               inno, iits, fp_out, ierr );
}

/***********************************************************************
 * Compute the inner source, i.e., the within-group scattering source.
 ***********************************************************************/
void inr_src ( input_data *input_vars, sn_data *sn_vars,
               control_data *control_vars, solvar_data *solvar_vars, para_data *para_vars )
{
/***********************************************************************
 * Local variables
 ***********************************************************************/
    int g;

/***********************************************************************
 * Compute the within-group scattering source. Thread over groups.
 ***********************************************************************/
#pragma omp parallel for schedule(dynamic, 1) default(shared) private(g)
    for ( g = 1; g <= NG; g++ )
    {
        if ( !(INRDONE_1D(g-1)) )
        {
            inr_src_scat ( input_vars, sn_vars, solvar_vars, g, para_vars );
        }
    }
}

/***********************************************************************
 * Compute the within-group scattering for a given group. Add it to fixed
 * and out-of-group sources.
 ***********************************************************************/
void inr_src_scat ( input_data *input_vars, sn_data *sn_vars,
                    solvar_data *solvar_vars, int g, para_data *para_vars )
{
/***********************************************************************
 * Local variables
 ***********************************************************************/
    int i, j, k, l, m, mom;

    double F_CS_temp[NX*NY*NZ];

/***********************************************************************
 * Loop over all cells. Set the first source moment with flux (f). Then
 * set remaining source moments with fluxm (fm) and combination of
 * higher scattering orders.
 ***********************************************************************/
#ifdef USEVML_unchecked
    if (NX*NY*NZ > VECLEN_MIN)
    {
        vmdMul(NX*NY*NZ, &F_3D(0,0,0), &CS_4D(0,0,0,0), F_CS_temp,
               VML_ACCURACY | VML_HANDLING | VML_ERROR );

        vmdAdd(NX*NY*NZ, F_CS_temp, &QO_4D(0,0,0,0), &Q_4D(0,0,0,0),
               VML_ACCURACY | VML_HANDLING | VML_ERROR );

        mom = 2;

        for ( l = 2; l <= NMOM; l++ )
        {
            for ( m = 1; m <= LMA_1D(l-1); m++ )
            {
                vmdMul(NX*NY*NZ, &FM_4D((mom-2),0,0,0),
                       &CS_4D((l-1),0,0,0), &Q_4D((mom-1),0,0,0),
                       VML_ACCURACY | VML_HANDLING | VML_ERROR );

                vmdAdd(NX*NY*NZ, &QO_4D((mom-1),0,0,0),
                       &Q_4D((mom-1),0,0,0), &Q_4D((mom-1),0,0,0),
                       VML_ACCURACY | VML_HANDLING | VML_ERROR );

                mom += 1;
            }
        }
    }

    else
    {
        for ( k = 1; k <= NZ; k++ )
        {
            for ( j = 1; j <= NY; j++ )
            {
                for ( i = 1; i <= NX; i++ )
                {
                    Q_4D(0,(i-1),(j-1),(k-1))
                        = QO_4D(0,(i-1),(j-1),(k-1))
                        + CS_4D(0,(i-1),(j-1),(k-1))
                        *F_3D((i-1),(j-1),(k-1));
                    mom = 2;

                    for ( l = 2; l <= NMOM; l++ )
                    {
                        for ( m = 1; m <= LMA_1D(l-1); m++ )
                        {
                            Q_4D((mom-1),(i-1),(j-1),(k-1))
                                = QO_4D((mom-1),(i-1),(j-1),(k-1))
                                + CS_4D((l-1),(i-1),(j-1),(k-1))
                                *FM_4D((mom-2),(i-1),(j-1),(k-1));

                            mom += 1;
                        }
                    }
                }
            }
        }
    }

#else
    for ( k = 1; k <= NZ; k++ )
    {
        for ( j = 1; j <= NY; j++ )
        {
            for ( i = 1; i <= NX; i++ )
            {
                Q_4D(0,(i-1),(j-1),(k-1))
                    = QO_4D(0,(i-1),(j-1),(k-1))
                    + CS_4D(0,(i-1),(j-1),(k-1))
                    *F_3D((i-1),(j-1),(k-1));

/***********************************************************************
 * Work on other moments with fluxm array
 ***********************************************************************/
                mom = 2;

                for ( l = 2; l <= NMOM; l++ )
                {
                    for ( m = 1; m <= LMA_1D(l-1); m++ )
                    {
                        Q_4D((mom-1),(i-1),(j-1),(k-1))
                            = QO_4D((mom-1),(i-1),(j-1),(k-1))
                            + CS_4D((l-1),(i-1),(j-1),(k-1))
                            *FM_4D((mom-2),(i-1),(j-1),(k-1));

                        mom += 1;
                    }
                }
            }
        }
    }
#endif
}

/***********************************************************************
 * Check for inner iteration convergence using the flux array.
 ***********************************************************************/
void inr_conv ( input_data *input_vars, para_data *para_vars,
                control_data *control_vars, solvar_data *solvar_vars,
                int inno, int *iits, FILE *fp_out, int *ierr )
{
/***********************************************************************
 * Local variables
 ***********************************************************************/
    int g, k, j, i;
    double dfmxi_tmp = -1;

    double df[NX*NY*NZ*NG];

/***********************************************************************
 * Thread group loops for computing local difference (df) array.
 * compute max for that group.
 ***********************************************************************/
// TODO: Add USEMKL
#pragma omp parallel for schedule(dynamic, 1) default(shared) \
    private(g)
    for ( g = 1; g <= NG; g++ )
    {
        if ( !INRDONE_1D(g-1) )
        {
            iits[g-1] = inno;

            for ( k = 0; k < NZ; k++ )
            {
                for ( j = 0; j < NY; j++ )
                {
                    for ( i = 0; i < NX; i++ )
                    {
                        if ( fabs(FLUXPI_4D(i,j,k,(g-1))) > TOLR )
                        {
                            DF_4D(i,j,k,(g-1))
                                = fabs( FLUX_4D(i,j,k,(g-1)) / FLUXPI_4D(i,j,k,(g-1)) - 1 );
                        }
                        else
                        {
                            DF_4D(i,j,k,(g-1))
                                = fabs( FLUX_4D(i,j,k,(g-1)) - FLUXPI_4D(i,j,k,(g-1)) );
                        }

                        dfmxi_tmp = MAX(dfmxi_tmp, DF_4D(i,j,k,(g-1)));
                    }
                }
            }

            DFMXI_1D(g-1) = dfmxi_tmp;
        }
    }

/***********************************************************************
 * All procs then reduce dfmxi for all groups, determine which groups
 * are converged and print requested info
 ***********************************************************************/
    glmax_d_1d(DFMXI, NG, COMM_SNAP);

    for ( g = 0; g < NG; g++ )
    {
        if ( DFMXI_1D(g) <= EPSI )
        {
            INRDONE_1D(g) = true;
        }
    }

    if ( (IPROC == ROOT) && (IT_DET == 1) )
    {
        for ( g = 1; g <= NG; g++ )
        {
            fprintf(fp_out, "    Group %i     Inner %i     Dfmxi %.4E\n",
                    g, iits[g-1], DFMXI_1D(g-1));
        }
    }
}
